core/shared/src/main/scala/coursier/util/Print.scala

```diff
@@ -146,81 +179,85 @@ object Print {
-              case (mod, ver) =>
-                Elem(
-                  Dependency(mod, ver, "", Set.empty, Attributes("", ""), false, false),
-                  excluded = true
-                )
-            }
+            case (mod, ver) =>
+              ElemImpl(
+                Dependency(mod, ver, "", Set.empty, Attributes("", ""), false, false),
+                excluded = true
+              )
+          }
 
-          dependencies.map(Elem(_, excluded = false)) ++
-            (if (printExclusions) excluded else Nil)
+          dependencies.map(ElemImpl(_, excluded = false)) ++
+            (if (withExclusions) excluded else Nil)
         }
     }
 
-    if (reverse) {
+    a => ElemImpl(a, excluded = false)
+  }
 
-      final case class Parent(
-        module: Module,
-        version: String,
-        dependsOn: Module,
-        wantVersion: String,
-        gotVersion: String,
-        excluding: Boolean
-      ) {
-        lazy val repr: String =
-          if (excluding)
-            s"$yellow(excluded by)$reset $module:$version"
-          else if (wantVersion == gotVersion)
-            s"$module:$version"
-          else {
-            val assumeCompatibleVersions = compatibleVersions(wantVersion, gotVersion)
-
-            s"$module:$version " +
-              (if (assumeCompatibleVersions) yellow else red) +
-              s"(wants $dependsOn:$wantVersion, got $gotVersion)" +
-              reset
-          }
-      }
+  def normalTree(roots: Seq[Dependency], resolution: Resolution, withExclusions: Boolean): Tree[Elem] = {
+    val elemFactory = getElemFactory(resolution, withExclusions)
+    Tree[Elem](roots.toVector.map(elemFactory), (elem: Elem) => elem.children)
+  }
+
+  def reverseTree(roots: Seq[Dependency], resolution: Resolution, withExclusions: Boolean): Tree[Parent] = {
+    val elemFactory = getElemFactory(resolution, withExclusions)
+
+    final case class ParentImpl(
+                                 module: Module,
+                                 version: String,
+                                 dependsOn: Module,
+                                 wantVersion: String,
+                                 gotVersion: String,
+                                 excluding: Boolean
+                               ) extends Parent {
+      def repr(colors: Colors): String =
+        if (excluding)
+          s"${colors.yellow}(excluded by)${colors.reset} $module:$version"
+        else if (wantVersion == gotVersion)
+          s"$module:$version"
+        else {
+          val assumeCompatibleVersions = compatibleVersions(wantVersion, gotVersion)
 
-      val parents: Map[Module, Seq[Parent]] = {
-        val links = for {
-          dep <- resolution.dependencies.toVector
-          elem <- Elem(dep, excluded = false).children
+          s"$module:$version " +
+            (if (assumeCompatibleVersions) colors.yellow else colors.red) +
+            s"(wants $dependsOn:$wantVersion, got $gotVersion)" +
```

Comment:
1. Could you keep the message consistent with normal tree? I think the `<wanted version> -> <reconciled version>` is good and succinct.

```
$ resolve -t org.apache.avro:avro:1.8.1 org.tukaani:xz:1.6
│  ├─ org.tukaani:xz:1.5 -> 1.6 (possible incompatibility)
```

2. The message appeared at the wrong place. 
```
$ resolve -T org.apache.avro:avro:1.8.1 org.tukaani:xz:1.6
├─ org.tukaani:xz:1.6 // should be here
│  └─ org.apache.avro:avro:1.8.1 (wants org.tukaani:xz:1.5, got 1.6)
```

Would be good to add a test case for this as well.

Reply:
Ok I realized that's the behavior on master. You can leave it as it is if you want. Thanks.